pulsar-zookeeper/src/main/java/com/yahoo/pulsar/zookeeper/ZooKeeperServerAspect.java

```diff
@@ -31,15 +31,26 @@
  */
 @Aspect
 public class ZooKeeperServerAspect {
+    private static boolean metricsRegistered = false;
+
     @Pointcut("execution(org.apache.zookeeper.server.ZooKeeperServer.new(..))")
-    public void processRequest() {
+    public void zkServerConstructorPointCut() {
     }
 
-    @After("processRequest()")
-    public void timedProcessRequest(JoinPoint joinPoint) throws Throwable {
+    @After("zkServerConstructorPointCut()")
+    public void zkServerConstructor(JoinPoint joinPoint) throws Throwable {
         // ZooKeeperServer instance was created
         ZooKeeperServer zkServer = (ZooKeeperServer) joinPoint.getThis();
 
+        synchronized (ZooKeeperServerAspect.class) {
+            if (metricsRegistered) {
```

Comment:
so, the variable is only read/written within the mutex and there is no need to use volatile, because the mutex implies here a full memory barrier, so all variables needs to be committed to memory and the execution cannot be reordered